### ISSUE TITLE
perf(server): apply structured-output token masks as a packed u32 bitmask instead of a full f32 bias

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1316]
## Summary

Every constrained-decoding step unpacks the grammar engine's packed token bitmask into a `Vec<bool>` the size of the vocabulary, turns that into a `Vec<f32>` bias of the same size, uploads the bias to the device and adds it to the logits. For a 248k-row head that is two 248k-element host loops plus a 1 MB host-to-device copy per token per constrained sequence, all on the scheduler thread between the forward pass and sampling. The change keeps the mask in its packed form (32 tokens per `u32`, 31 KB for the same vocabulary), uploads that, and expands it on the device with four graph ops feeding `where_cond`, removing both host loops and 97% of the copy.

## Current behavior

`src/server/structured.rs:345` (`compute_mask`) unpacks the packed mask the grammar engine returns:

```rust
self.mask_buf.resize(self.vocab_size, false);
...
vob.iter_set_entries(|idx| { if idx < vocab_size { mask[idx] = true; } });
```

`src/server/structured.rs:736` (`apply_structured_mask_to_logits`) then walks the whole vocabulary a second time to build the bias and ships it:

```rust
constraint.bias_buf.resize(vocab_size, 0.0);
for (i, slot) in constraint.bias_buf.iter_mut().enumerate() {
    let allowed = i < mask_len && constraint.mask_buf[i];
    if !allowed { *slot = f32::NEG_INFINITY; }
}
let bias_arr = mlxcel_core::from_slice_f32(&constraint.bias_buf, &[1, vocab_size as i32]);
Ok(mlxcel_core::add(logits, &bias_arr))
```

`vocab_size_hint` is the logits width (padded head rows, 248320 for Qwen3.8 against a 248077-token tokenizer), so the bias is `4 * 248320 = 993 KB` per step. The scheduler calls this inside `apply_structured_mask` (`src/server/batch/scheduler.rs:3131`) for each constrained sequence on each tick, before sampling, so with B concurrent constrained sequences the cost is B times per tick. The grammar engine's own mask computation is roughly 50 microseconds (`src/server/structured.rs:44`); the unpack, the bias fill and the copy are the part this change removes. The existing `mask_buf` / `bias_buf` reuse (`src/server/structured.rs:288` onward) avoids allocation but not the loops.

All device ops needed are already bound: `from_slice_u32` (`src/lib/mlxcel-core/src/lib.rs:126`), `take` (`:850`), `right_shift` (`:1014`), `bitwise_and` (`:1002`), `where_cond` (`:524`).

## Expected behavior

Per step, for a constraint whose matcher vocabulary is `V_tok` and a logits width `V`:

```text
words      = packed mask from the matcher, u32[ceil(V_tok / 32)]   (no unpack)
words_dev  = from_slice_u32(words padded with zeros to ceil(V / 32), [ceil(V/32)])
word_idx   = precomputed i32[V]  = i >> 5          (built once per constraint, cached on device)
bit_idx    = precomputed u32[V]  = i & 31          (built once per constraint, cached on device)
allowed    = bitwise_and(right_shift(take(words_dev, word_idx, 0), bit_idx), 1) != 0   -> bool[V]
masked     = where_cond(allowed, logits, neg_inf)  with neg_inf a cached [1] array in logits dtype
```

Padding rows past `V_tok` read zero words and are therefore disallowed, preserving the current rule that a token the sampler could emit but the matcher does not know is masked out. The empty-mask check (no allowed token within the first `V` entries) is computed on the packed words: every word up to `ceil(V/32)` is zero, with the last partial word masked to the valid bit range. Output is numerically identical to today's bias add for every allowed token (the logit is passed through unchanged rather than having `0.0` added) and identical for disallowed tokens (`-inf`).

The two cached index arrays are `2 * 4 * V` bytes (about 2 MB for V = 248320), allocated once per constraint and reused for its lifetime; the per-step upload drops from `4V` bytes to `ceil(V/32) * 4` bytes.

## Implementation plan

1. `src/server/structured.rs`:
   - Add to `StructuredOutputConstraint`: `packed_buf: Vec<u32>` (capacity `ceil(vocab_size_hint / 32)`), `word_idx: Option<UniquePtr<MlxArray>>`, `bit_idx: Option<UniquePtr<MlxArray>>`, `neg_inf: Option<UniquePtr<MlxArray>>`, keyed by `vocab_size_hint` and the logits dtype seen at first use (rebuilt if either changes).
   - Add `pub fn compute_packed_mask(&mut self) -> Result<&[u32], StructuredOutputError>` that calls `matcher.compute_mask_or_eos()` exactly as `compute_mask` does and copies `vob.as_slice()` (the packed `u32` words) into `packed_buf`, zero-padding to the hint width. Keep `compute_mask` (the `Vec<bool>` form) for the unit tests that inspect individual entries; it is no longer on the hot path.
   - Rewrite `apply_structured_mask_to_logits` to the graph above. `vocab_size_hint` semantics are unchanged.
2. `src/server/batch/scheduler.rs:3131` (`apply_structured_mask`): no signature change; it keeps calling `apply_structured_mask_to_logits`.
3. If profiling shows the four-op expansion itself dominates (it should not; the ops are elementwise over V and fuse), follow up with a single custom kernel `logits[i] = (words[i>>5] >> (i&31)) & 1 ? logits[i] : -inf` using the in-tree fused kernel pattern under `src/lib/mlx-cpp/turbo/` with a `{"T", logits.dtype()}` template key (the CUDA cache-key rule in `CLAUDE.md` applies). Not required for this issue.
4. `make verify-kernel-dtype-keys` stays green (no CUDA kernel is added in the graph-op version).

## Validation

(a) Unit tests in `src/server/structured_tests.rs`:

- `packed_mask_matches_bool_mask` (for a small schema and the stub tokenizer, `compute_packed_mask` bit `i` equals `compute_mask()[i]` for every `i < vocab_size`, and padding bits are zero).
- `packed_apply_matches_bias_apply` (keep the old bias routine as a test-only reference: for random logits of width `vocab_size_hint > vocab_size`, the new `apply_structured_mask_to_logits` output equals the reference elementwise, `-inf` positions included).
- `packed_apply_empty_mask_errors` (a stopped / empty matcher still returns the `Matcher` error).
- `packed_apply_rebuilds_index_arrays_on_width_change`.
- Existing `structured_tests.rs` end-to-end tests (schema conformance through the scheduler) must pass unchanged.

(b) Real checkpoint and benchmark: `mlx-community/Qwen3.8-27B-4bit` (248320-row head, the widest in the test set) and `mlx-community/Qwen3-4B-4bit`. Stop Time Machine before measuring (`tmutil stopbackup`).

```bash
./target/release/mlxcel-server -m models/Qwen3.8-27B-4bit --port 8080   # after: ./target/release/mlxcel download mlx-community/Qwen3.8-27B-4bit
# constrained
for i in 1 2 3; do curl -s localhost:8080/v1/chat/completions -H 'content-type: application/json' -d '{"model":"q","messages":[{"role":"user","content":"List 40 European capitals with their country and population as JSON."}],"max_tokens":512,"temperature":0,"response_format":{"type":"json_schema","json_schema":{"name":"caps","schema":{"type":"object","properties":{"items":{"type":"array","items":{"type":"object","properties":{"city":{"type":"string"},"country":{"type":"string"},"population":{"type":"integer"}},"required":["city","country","population"],"additionalProperties":false}}},"required":["items"],"additionalProperties":false}}}}' | jq '.usage.completion_tokens'; done
# unconstrained control: same request without response_format
```

Acceptance: the constrained output still validates against the schema (pipe through `jq` and a schema check) and is token-identical to the pre-change binary at `temperature: 0`; the constrained decode rate (completion tokens divided by the server-logged generation time, median of three) improves on the 27B by at least the per-token cost removed (expected 0.5 to 1.5 ms per token at 248k vocabulary, so a measurable gain at the 27B's 18 tok/s baseline) and does not regress on the 4B; the unconstrained control is unchanged.

## Acceptance criteria

- [x] The hot path uploads a packed `u32` mask and expands it on device; no per-token `Vec<bool>` or `Vec<f32>` walk over the vocabulary remains.
- [x] Padded head rows beyond the matcher vocabulary stay masked.
- [ ] Output is token-identical to the bias implementation at temperature 0 on both checkpoints.
- [ ] Benchmark shows no regression on the 4B and an improvement on the 27B, recorded in `docs/benchmarks.md` with the method above.
- [x] Unit tests above added.
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes
- [ ] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass

## Out of scope

- A fused masked-argmax sampler for greedy constrained decoding.
- Sharing one mask upload across sequences that hold the same grammar state.

**Comment by inureyes** [comment]
Implemented in #1578. The packed path landed with one deviation from the plan in this issue, and four acceptance boxes are deliberately still open.

Deviation: the plan cached a `word_idx` `i32[V]` and a `bit_idx` `u32[V]` per constraint and gathered with `take`. The implementation instead broadcasts, which needs no per-constraint index tables at all: the packed words go up as a `u32[n_words, 1]` column and the 32 bit positions as a `u32[1, 32]` row, so one `right_shift` plus a `bitwise_and` with 1 produces `[n_words, 32]` whose row-major order is already token id `w * 32 + b`. A reshape, a trim to the logits width and a cast give the same `bool[V]` the plan built. That drops the 2 MB of cached index arrays per constrained sequence, removes the gather of `V` indices from the per-step graph, and makes the rebuild-on-width-change case in the plan's fourth test unnecessary, so that test became `packed_apply_handles_a_width_change_between_calls` instead. It is also the expansion MLX itself uses to unpack non-power-of-two quantized weights, so it is a well-trodden shape. Everything else follows the plan: `compute_packed_mask` was added and `compute_mask` was kept as the `Vec<bool>` accessor, `apply_structured_mask_to_logits` kept its signature, and `src/server/batch/scheduler/run_loop.rs` is untouched.

Measured at the Qwen3.8-27B geometry (248320 logits rows against 248077 tokenizer entries), min-of-60 per trial, median of four runs on an M1 Ultra: the scheduler-thread cost of preparing and uploading the mask falls from 711 us to 7.4 us per step per constrained sequence, and the same work followed by a forced eval falls from 1002 us to 322 us. The reproduce command and the full table are in the PR body.

Still open, and why: the two real-checkpoint boxes (token-identical output at temperature 0, and the `docs/benchmarks.md` entry) need a live server against `mlx-community/Qwen3.8-27B-4bit` and `mlx-community/Qwen3-4B-4bit`, which this run did not have; the PR body carries the exact request to run. The two whole-workspace gate boxes are unchecked because this run was scoped to narrow commands: `--lib server::structured` (21 passed), `--test structured_outputs` (21 passed, the full pre-existing suite unchanged), `clippy --profile test-fast --lib --tests --features metal,accelerate -- -D warnings` (clean) and `cargo fmt --all -- --check` (clean). The workspace-wide `cargo test` and `cargo clippy --workspace --all-targets` remain for whoever runs the full gate.
